### PR TITLE
xds: gracefully handle stream termination without error

### PR DESCRIPTION
**Pull request by howardjohn**
Today, we only reconnect on the stream erroring. If it ends with a
success, we ~infinite loop doing nothing.

In standard operations, Istiod will never end the stream without an
error. However, it is possible to expose Istiod with a proxy that does.

### CODE REVIEW

**Comment by howardjohn** [comment]
Reproducer:

<details>

```yaml
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 9901 }

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address: { address: 127.0.0.1, port_value: 15011 }
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          stream_idle_timeout: 2s
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route: { cluster: some_service }
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
  - name: some_service
    connect_timeout: 0.25s
    type: STATIC
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: some_service
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 127.0.0.1
                port_value: 15010

    typed_extension_protocol_options:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        explicit_http_config:
          http2_protocol_options:
            connection_keepalive:
              interval: 30s
              timeout: 5s
```

`PROXY_WORKLOAD_INFO=default/local/default PROXY_MODE=dedicated FAKE_CA=true XDS_ADDRESS="http://localhost:15011" cargo run -F testing`
</details>

File: src/xds/client.rs [hunk 1]
```diff
@@ -647,10 +645,14 @@ impl AdsClient {
                     self.handle_demand_event(_demand_event, &discovery_req_tx).await?;
                 }
                 msg = response_stream.message() => {
-                    let msg = msg?;
+                    let Some(msg) = msg? else {
+                        // If we got a None message, the stream ended without error.
+                        // This could be an explicit OK response, or if the stream is reset without a gRPC status.
```
Comment (keithmattix) [inline src/xds/client.rs:650]: Should we log so that we know this was a stream termination?

Comment (howardjohn) [reply src/xds/client.rs:650]: It is logged above, you will see logs like:

```
2024-11-22T20:37:36.545633Z     warn    xds::client:xds{id=7}   XDS client complete
2024-11-22T20:37:36.552650Z     info    xds::client:xds{id=8}   Stream established
2024-11-22T20:37:36.553004Z     info    xds::client:xds{id=8}   received response       type_url="type.googleapis.com/istio.workload.Address" size=40 removes=0
2024-11-22T20:37:36.554415Z     info    xds::client:xds{id=8}   received response       type_url="type.googleapis.com/istio.security.Authorization" size=0 removes=0
2024-11-22T20:37:38.555672Z     warn    xds::client:xds{id=8}   XDS client complete
2024-11-22T20:37:38.568035Z     info    xds::client:xds{id=9}   Stream established
2024-11-22T20:37:38.568344Z     info    xds::client:xds{id=9}   received response       type_url="type.googleapis.com/istio.workload.Address" size=40 removes=0
2024-11-22T20:37:38.570170Z     info    xds::client:xds{id=9}   received response       type_url="type.googleapis.com/istio.security.Authorization" size=0 removes=0
```

(test has the server close it every 2s)

Comment (keithmattix) [reply src/xds/client.rs:650]: Ah you mean the `XDS client complete` message? That's probably sufficient.

File: src/xds/client.rs [hunk 2]
```diff
@@ -647,10 +645,14 @@ impl AdsClient {
...
+                        return Ok(());
+                    };
                     let mut received_type = None;
                     if !self.types_to_expect.is_empty() {
-                        received_type = msg.as_ref().map(|e| e.type_url.clone());
+                        received_type = Some(msg.type_url.clone())
                     }
                     if let XdsSignal::Ack = self.handle_stream_event(msg, &discovery_req_tx).await? {
```
Comment (keithmattix) [inline src/xds/client.rs:657]: Probably not related to this PR, but why isn't there any Nack handling?

**Review by bleggett — APPROVED** [review]

**Pull request merged**